### ISSUE TITLE
Issue with rust-toolchain file

### ISSUE THREAD

**Issue by WyseNynja** [trailofbits/siderophile#14]
My project has a rust-toolchain file in its root for a nightly toolchain, but `./setup.sh` simply calls "cargo" so it built with a different version than my project. This lead to the following error when running analyze.sh.

```
$ ~/src/siderophile/analyze.sh myapp
trawling source code of dependencies for unsafety
   Compiling libc v0.2.60

...


error[E0514]: found crate `cached` compiled by an incompatible version of rustc myapp                                   
  --> src/lib.rs:15:1
   |
15 | extern crate cached;
   | ^^^^^^^^^^^^^^^^^^^^
   |
   = help: please recompile that crate using this compiler (rustc 1.37.0-nightly (8aa42ed7c 2019-06-24))
   = note: the following crate versions were found:
           crate `cached` compiled by rustc 1.36.0 (a53f9df32 2019-07-03): ~/src/myapp/target/debug/deps/libcached-b5e2b19b56d13d52.rmeta

error: aborting due to previous error

thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Cargo("Could not compile `myapp`.")', src/libcore/result.rs:1051:5
stack backtrace:
   0: std::panicking::default_hook::{{closure}}
   1: std::panicking::default_hook
   2: std::panicking::rust_panic_with_hook
   3: std::panicking::continue_panic_fmt
   4: rust_begin_unwind
   5: core::panicking::panic_fmt
   6: core::result::unwrap_failed
   7: siderophile::main
   8: std::rt::lang_start::{{closure}}
   9: std::panicking::try::do_call
  10: __rust_maybe_catch_panic
  11: std::rt::lang_start_internal
  12: main
```

To fix this problem, I had to build and run siderophile with the correct toolchain selected.

```
cd ~/src/siderophile
RUSTUP_TOOLCHAIN=$(cat ~/src/myapp/rust-toolchain) cargo build --release
```

```
$ cd ~/src/myapp
$ RUSTUP_TOOLCHAIN=$(cat rust-toolchain) ~/src/siderophile/analyze.sh myapp

trawling source code of dependencies for unsafety
   Compiling libc v0.2.60
...
generating LLVM bitcode for the callgraph
...
```

A note about this should probably be added to the readme.

**Comment by WyseNynja** [comment]
Well I got another error after it ran for a while. Do you think this is related, or a seperate issue?

```
$ RUSTUP_TOOLCHAIN=$(cat rust-toolchain) ~/src/siderophile/analyze.sh myapp

trawling source code of dependencies for unsafety
...
    Finished dev [unoptimized + debuginfo] target(s) in 3m 10s
generating LLVM bitcode for the callgraph
...
    Finished dev [unoptimized + debuginfo] target(s) in 13m 57s
generating callgraph
Cannot find LLVM bitcode in ./target/debug/deps or ../target/debug/deps
```

```
$  ls siderophile_out/
unsafe_deps.txt
```

**Comment by rozbb** [comment]
Noted. I actually ran into this issue myself on a project. I will update the README.

Do you think siderophile should respect `rust-toolchain`? If so, do you have any suggestions on how to make that work, since the override works on the `rustup` level?

**Comment by disconnect3d** [comment]
Just to give some more context. The `rust-toolchain` overrides the toolchain version that will be used for compiling; this is documented here: https://doc.rust-lang.org/stable/edition-guide/rust-2018/rustup-for-managing-rust-versions.html#managing-versions.

However, when you create a crate `A` which then uses an external crate `B` which has the `rust-toolchain` file with `X` toolchain specified, Cargo will not use `X`. It will use the default toolchain `A` is compiled with, also for compiling `B`. Otherwise, we would have (abi?) conflicts between the two.

The problem here is that Siderophile somehow respects `B` crate's `rust-toolchain` file and tries to compile it with `X` and then we have a clash.

I prepared a [failing] PoC in https://github.com/trailofbits/siderophile/pull/22 and I am trying to track down why this occurs and how to fix that. It seems that the compilation of `B` fails on https://github.com/trailofbits/siderophile/blob/master/src/deps.rs#L434 as in:
```
➜  crate-use-rust-toolchain git:(add-rust-toolchain) ✗ ../../target/debug/siderophile -o ./siderophile_out/unsafe_deps.txt
    Checking bitvec v0.15.2
error: 'rustc' is not installed for the toolchain '1.39.0-x86_64-apple-darwin': bitvec
To install, run `rustup component add rustc --toolchain 1.39.0-x86_64-apple-darwin`
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Cargo("Could not compile `bitvec`.")', src/libcore/result.rs:1187:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

**Comment by disconnect3d** [comment]
An obvious workaround is of course explicitly using the `RUSTUP_TOOLCHAIN` environment variable:
```
➜  crate-use-rust-toolchain git:(add-rust-toolchain) ✗ RUSTUP_TOOLCHAIN='nightly-x86_64-apple-darwin' ../../analyze.sh crate-uses-rust-toolchain
trawling source code of dependencies for unsafety
    Checking bitvec v0.15.2
    Checking crate-uses-rust-toolchain v0.1.0 (/Users/dc/tob/projects/siderophile/tests/crate-use-rust-toolchain)
    Finished dev [unoptimized + debuginfo] target(s) in 2.34s
generating LLVM bitcode for the callgraph
   Compiling bitvec v0.15.2
   Compiling crate-uses-rust-toolchain v0.1.0 (/Users/dc/tob/projects/siderophile/tests/crate-use-rust-toolchain)
    Finished dev [unoptimized + debuginfo] target(s) in 2.56s
generating callgraph
WARNING: You're attempting to print out a bitcode file.
This is inadvisable as it may cause display problems. If
you REALLY want to taste LLVM bitcode first-hand, you
can force output with the `-f' option.

Writing 'callgraph.dot'...
unmangling callgraph symbols
matching unsafe deps with callgraph nodes
tracing the unsafety up the tree
done. see siderophile_out/badness.txt
```

**Comment by disconnect3d** [comment]
Fixed via a workaround in #22.

I opened #25 that explains future work on that matter.
